meterfs/crypt.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Meterfs cryptography
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Aleksander Kaminski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdlib.h>
+#include <errno.h>
+#include <string.h>
+
+#include "tinyaes/aes.h"
+#include "crypt.h"
+
+
+static unsigned int meterfs_hash(unsigned int seed)
+{
+	int result;
+
+	seed *= 1103515245;
+	seed += 12345;
+	result = (seed >> 16) & 0x7ff;
+
+	seed *= 1103515245;
+	seed += 12345;
+	result <<= 10;
+	result ^= (seed >> 16) & 0x3ff;
+
+	seed *= 1103515245;
+	seed += 12345;
+	result <<= 10;
+	result ^= (seed >> 16) & 0x3ff;
+
+	return result;
+}
+
+
+void meterfs_constructIV(uint8_t *iv, const file_t *f, const entry_t *e)
+{
+	static const uint8_t filler[16] = {
+		0x55, 0xaa, 0x55, 0xaa, 0x55, 0xaa, 0x55, 0xaa,
+		0x55, 0xaa, 0x55, 0xaa, 0x55, 0xaa, 0x55, 0xaa
+	};
+	unsigned liv[4] = {
+		e->id.no,
+		meterfs_hash(e->id.no),
```

Comment:
This does not need to be hashed. IV is not intended to provide entropy. This can only lower security as hash can introduce collisions.

Filler can be initialized with 0.

Reply:
There won't be iv collision because of it, as I still provide unhashed unique id and sector combo too.
Problem is, I can only get like 2-3 unique bytes per IV and I tried to boots it's value by hashing.

If there's no point, I can remove it, less code, less computing power needed.

Reply:
I missed that hashed values are included too. This does not decrease security.

Additional value should be added to ensure IV is changed after file resize (including resize same size).

Reply:
Ok, how about then:
- record id,
- sector,
- file size,
- record size

and hash being removed? 16 MSB bytes are for future proofing, as we use AES128 for now, so doesn't matter. I can zero them if there's benefit over 0x55 0xaa pattern

Reply:
You need to leave enough lower bits to ensure block count does not overlap with unique bits. It is common practice to resize with same file size / record size to clear content. This means some other value needs to be added.

Reply:
I added file id (uid), which is unique for each file reincarnation (i.e. it is incremented on each file erase caused by either file resize or enabling/disabling encryption) and it is used in IV construction